File: src/services/reviewGenerator.ts
/**
 * 好评文案生成服务
 * 负责调用 DeepSeek API 生成买家好评文案
 */

import type { DeepSeekMessage, DeepSeekConfig } from "../config/types";
import type { ParsedGoodsInfo } from "./goodsParser";
import { DeepSeekClient, createDeepSeekClient } from "../api/deepseek";
import { Logger } from "../utils/logger";
import { saveTextFile, getReviewFilePath, formatReviewContent } from "../utils/file";

/**
 * 错误类：好评文案生成相关错误
 */
export class ReviewGeneratorError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ReviewGeneratorError";
  }
}

/**
 * 好评文案生成服务
 */
export class ReviewGenerator {
  private client: DeepSeekClient;
  private config: DeepSeekConfig;
  private logger: Logger;

  /**
   * 创建好评文案生成服务
   * @param config - DeepSeek 配置
   * @param logger - 日志工具
   */
  constructor(config: DeepSeekConfig, logger: Logger) {
    this.client = createDeepSeekClient({
      url: config.url,
      apiKey: config.apiKey,
    });
    this.config = config;
    this.logger = logger;
  }

  /**
   * 生成好评文案
   * @param goodsInfo - 解析后的商品信息
   * @param targetDir - 目标目录
   * @returns 生成的好评文案数组
   */
  public async generate(goodsInfo: ParsedGoodsInfo, targetDir: string): Promise<string[]> {
    const reviews: string[] = [];
    const reviewCount = this.config.reviewCount;

    this.logger.info(`开始生成好评文案，共 ${reviewCount} 条`);

    for (let i = 0; i < reviewCount; i++) {
      this.logger.info(`生成第 ${i + 1} 条好评文案...`);

      try {
        const review = await this.generateSingleReview(goodsInfo, i);
        reviews.push(review);
        this.logger.info(`第 ${i + 1} 条好评文案生成成功`);
      } catch (error) {
        this.logger.error(`第 ${i + 1} 条好评文案生成失败`, error as Error);
        // 继续生成下一条，不中断流程
      }

      // 每次生成之间添加延迟，避免 API 限流
      if (i < reviewCount - 1) {
        await this.delay(1000);
      }
    }

    if (reviews.length === 0) {
      throw new ReviewGeneratorError("所有好评文案生成失败");
    }

    // 保存到文件
    const reviewFilePath = getReviewFilePath(targetDir);
    const content = formatReviewContent(reviews);
    saveTextFile(reviewFilePath, content);

    this.logger.info(`好评文案已保存到: ${reviewFilePath}`);

    return reviews;
  }

  /**
   * 生成单条好评文案
   * @param goodsInfo - 解析后的商品信息
   * @param index - 当前生成索引
   * @returns 生成的好评文案
   */
  private async generateSingleReview(goodsInfo: ParsedGoodsInfo, _index: number): Promise<string> {
    // 构建提示词
    const messages = this.buildPrompt(goodsInfo);

    // 使用不同的 temperature 增加多样性
    const temperature = 0.7 + Math.random() * 0.3; // 0.7 - 1.0 之间随机

    const content = await this.client.chat(messages, {
      temperature,
      maxTokens: 500,
    });

    return this.cleanReviewContent(content);
  }

  /**
   * 构建提示词
   * @param goodsInfo - 解析后的商品信息
   * @returns 消息列表
   */
  private buildPrompt(goodsInfo: ParsedGoodsInfo): DeepSeekMessage[] {
    const userContent = [
      this.config.systemPrompt,
      "",
      "商品信息：",
      this.formatGoodsInfo(goodsInfo),
      "",
      "请生成80字左右的买家好评文案",
    ].join("\n");

    return [
      { role: "system", content: this.config.systemPrompt },
      { role: "user", content: userContent },
    ];
  }

  /**
   * 格式化商品信息
   * @param goodsInfo - 解析后的商品信息
   * @returns 格式化的商品信息文本
   */
  private formatGoodsInfo(goodsInfo: ParsedGoodsInfo): string {
    const lines: string[] = [
      `商品名称: ${goodsInfo.name}`,
      `品类: ${goodsInfo.category}`,
      `适用性别: ${goodsInfo.gender}`,
      `适用年龄: ${goodsInfo.ageGroup}`,
      `年份: ${goodsInfo.year}`,
      `季节: ${goodsInfo.season}`,
    ];

    if (goodsInfo.pantsLength) {
      lines.push(`裤长: ${goodsInfo.pantsLength}`);
    }

    lines.push(`袖长: ${goodsInfo.sleeveLength}`);
    lines.push(`领型: ${goodsInfo.collarType}`);
    lines.push(`穿戴方式: ${goodsInfo.wearType}`);
    lines.push(`特殊卖点: ${goodsInfo.features.join("、")}`);

    return lines.join("\n");
  }

  /**
   * 清理好评文案内容
   * @param content - 原始内容
   * @returns 清理后的内容
   */
  private cleanReviewContent(content: string): string {
    // 移除可能的引号
    let cleaned = content.replace(/^["']|["']$/g, "");

    // 移除可能的序号标记
    cleaned = cleaned.replace(/^\d+[\.\、]\s*/gm, "");

    return cleaned.trim();
  }

  /**
   * 延迟函数
   * @param ms - 延迟毫秒数
   */
  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

/**
 * 创建好评文案生成服务的工厂函数
 * @param config - DeepSeek 配置
 * @param logger - 日志工具
 * @returns ReviewGenerator 实例
 */
export function createReviewGenerator(config: DeepSeekConfig, logger: Logger): ReviewGenerator {
  return new ReviewGenerator(config, logger);
}

File: src/services/reviewGenerator.py
"""好评文案生成服务模块"""
import os
import sys
import time
from typing import List
from src.api.deepseek import DeepSeekClient
from src.config.types import Config, GoodsData
from src.services.goodsParser import GoodsParser
from src.utils.file import save_text_file


class ReviewGenerator:
    """好评文案生成服务"""
    
    def __init__(self, config: Config):
        self.config = config
        self.client = DeepSeekClient(
            api_key=config.deepseek.apiKey,
            model=config.deepseek.model,
            base_url=config.deepseek.url
        )
        self.parser = GoodsParser()
    
    def generate_reviews(self, goods_data: GoodsData, target_dir: str) -> List[str]:
        """生成好评文案并保存到文件"""
        # 构建商品信息字符串
        goods_info_str = self.parser.parse_goods_info(goods_data.goods)
        
        # 获取系统提示词
        system_prompt = self.config.deepseek.systemPrompt
        
        # 获取生成次数
        review_count = self.config.deepseek.reviewCount
        
        print(f"\n📝 开始生成好评文案: 共 {review_count} 条")
        print("-" * 50)
        
        # 生成多条好评文案（带进度展示）
        reviews = self._generate_reviews_with_progress(
            system_prompt=system_prompt,
            goods_info=goods_info_str,
            count=review_count
        )
        
        if not reviews:
            print("\n⚠️ 未能生成任何好评文案")
            return []
        
        print(f"\n✅ 好评文案生成完成: 成功 {len(reviews)}/{review_count} 条")
        
        # 保存到文件
        self._save_reviews(reviews, target_dir)
        
        return reviews
    
    def _generate_reviews_with_progress(self, system_prompt: str, goods_info: str, count: int) -> List[str]:
        """生成多条好评文案并显示进度"""
        reviews = []
        start_time = time.time()
        
        for i in range(count):
            # 每次使用不同的 temperature 值
            temp = 0.7 + (1.0 - 0.7) * (i / max(count - 1, 1))
            
            # 显示当前进度
            elapsed = time.time() - start_time
            percent = int(((i + 1) / count) * 100)
            bar_length = 20
            filled = int((i + 1) / count * bar_length)
            bar = "█" * filled + "░" * (bar_length - filled)
            
            print(f"\r  📝 好评文案进度: [{bar}] {percent}% ({i+1}/{count})", end="")
            sys.stdout.flush()
            
            # 生成单条好评文案
            review = self.client.generate_review(system_prompt, goods_info, temperature=temp)
            if review:
                reviews.append(review)
                print(f"\r  ✅ 第 {i+1} 条好评文案生成成功")
            else:
                print(f"\r  ❌ 第 {i+1} 条好评文案生成失败")
            
            # 延迟等待，避免 API 限流
            if i < count - 1:
                time.sleep(1)
        
        print()  # 换行
        return reviews
    
    def _save_reviews(self, reviews: List[str], target_dir: str) -> None:
        """保存好评文案到文件"""
        content_lines = []
        
        for i, review in enumerate(reviews, 1):
            content_lines.append(f"=== 第{i}条 ===")
            content_lines.append(review.strip())
            content_lines.append("")  # 空行
        
        content = "\n".join(content_lines)
        
        output_path = os.path.join(target_dir, "review.txt")
        save_text_file(output_path, content)
        
        print(f"好评文案已保存到: {output_path}")